FEAT: Add OAUTH2 support for POP3 polling

AlessioC31 opened the pull request:
Hello, I've worked on the implementation of OAUTH2 protocol for POP3 polling.

It depends on this change on the `net-pop` library: https://github.com/ruby/net-pop/pull/16 . As you can see it is not yet merged, but I'm opening already this PR so hopefully we could start iterating on it (and maybe you know someone who could review the PR in the `net-pop` repo?)

By the way, this is how this change works:

The users configures it via these settings in site settings:

- `pop3_polling_oauth2_endpoint`: The endpoint to use to get the access token
- `pop3_polling_oauth2_clientid`: The clientid to use to get the access token. This is the client id of the delegated application registered in the oauth2 tenant
- `pop3_polling_oauth2_refresh_token`: The initial refresh token, used to get the access token
- `pop3_polling_oauth2_scope`: The scope to use to get the access token, it should contain both `POP` (this varies across every service, for outlook it is `https://outlook.office.com/POP.AccessAsUser.All`) and `offline_access` in order to get a new refresh token each time.

Discourse then uses this configuration to get the `access_token` and its `expire_at`, plus a new `refresh_token`. It stores the `access_token` in redis with expiration `expire_at` and it saves the new `refresh_token` back in the site settings (this I'm not sure if it's the best way to do it).
Every time the `PollMailbox` job is executed, it first checks if the redis key has expired, in this case it refreshes the access token and then uses it to access the mailbox. If it has not expired, it use the value stored in redis as access token.

Please let me know if I was not clear enough and if you think something should be changed.

Thanks!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/discourse/discourse?pullRequest=20445) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/discourse/discourse?pullRequest=20445) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/discourse/discourse?pullRequest=20445) it.</sub>

SamSaffron commented on spec/jobs/poll_mailbox_spec.rb:
```diff
@@ -99,6 +99,51 @@
+          1.seconds,
+          "old_access",
+        )
+        sleep 1
```
this is risky in a test, it will end up flaking, can it be done in another way?

AlessioC31 commented:
Hello, I saw that there were some failing checks. I fixed the ones regarding `SiteSettingPlugin`, it was because the new properties in `site_settings.yml` were empty by default, I changed them to `""`.
There will still be failing checks since the `OAUTH2` PR has not been merged into `net-pop`

AlessioC31 commented:
Hello, since this seems blocked on the `net-pop` part (I will add more tests and documentation as suggested). We need to think on how to move forward on this. Would it be fine for you if instead of supporting `OAUTH2`, we add support only for Microsoft Exchange? We could do it by using the Microsoft Graph API (https://learn.microsoft.com/en-us/graph/api/resources/mail-api-overview?view=graph-rest-1.0, sdk: https://github.com/microsoftgraph/msgraph-sdk-ruby)

Thanks!

cc @SamSaffron

nattsw commented:
@AlessioC31 I'm going to close this PR. But when the upstream dependency has merged, please feel free to re-open. 🤗

nattsw commented:
@AlessioC31 I'm going to close this PR. But when the upstream dependency has merged, please feel free to re-open. 🤗
